## Importing Libraries

In [1]:
# Importing Libraries

import pandas as pd
import numpy as np
import os

## Importing Dataframes

In [2]:
# Define path
path = r'/Users/isaacotubanjo/Documents/08:08:2023 Instacart Business Analysis'

In [3]:
# Import orders_products_aggregated.pkl as df_ords_prods
df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [4]:
# Import customers.csv as df_customers
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

## Wrangling Data

In [5]:
# Conducting preliminary checks on dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df_customers.shape

(206209, 10)

In [7]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
df_ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...    price_range_loc  busiest_day_of_week  \
0         Soda  ...  Mid-range product       Regularly busy   
1         Soda  ...  Mid-range product       Regularly busy   
2         Soda  ...  Mid-range product       Regularly busy   
3         Soda  ...  Mid-range product           Least busy   
4         Soda  ...  Mid-range product           Least busy   

         busiest_days busiest_period_of_day max_order  loyalty_flag avg_price  \
0  Regularly busy day        Average orders        10  New customer  6.367797   
1     Least busy days        Average orders        10  New customer  6.367797   
2     Least busy days           Most orders        10  New customer  6.367797   
3     Least busy days        Average orders        10  New customer  6.367797   
4     Least busy days           Most orders        10  New customer  6.367797   

   spending_flag median_days   order_frequency_flag  
0    Low spender        20.5  Non-frequent customer  
1    Low spender        20.5  Non-frequent customer  
2    Low spender        20.5  Non-frequent customer  
3    Low spender        20.5  Non-frequent customer  
4    Low spender        20.5  Non-frequent customer  

[5 rows x 23 columns]

In [9]:
df_ords_prods.shape

(32404859, 23)

In [10]:
# Checking for Missing or NaN values in all columns in the dataframe
df_customers['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [11]:
df_customers['STATE'].value_counts(dropna = False)

Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana       

In [12]:
df_customers['Age'].value_counts(dropna = False)

19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: Age, Length: 64, dtype: int64

In [13]:
df_customers['n_dependants'].value_counts(dropna = False)

0    51602
3    51594
1    51531
2    51482
Name: n_dependants, dtype: int64

In [14]:
df_customers['fam_status'].value_counts(dropna = False)

married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: fam_status, dtype: int64

In [15]:
df_customers['income'].value_counts(dropna = False)

57192     10
95891     10
95710     10
97532      9
98675      9
          ..
73141      1
71524      1
74408      1
44780      1
148828     1
Name: income, Length: 108012, dtype: int64

There seems to be no missing values in the dataframe

In [17]:
# Renaming columns 
df_customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [18]:
df_customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [19]:
df_customers.rename(columns = {'date_joined' : 'Date Joined'}, inplace = True)

In [20]:
df_customers.rename(columns = {'n_dependants' : 'No of dependants'}, inplace = True)

In [21]:
df_customers.rename(columns = {'fam_status' : 'Marital Status'}, inplace = True)

In [22]:
df_customers.rename(columns = {'income' : 'Income'}, inplace = True)

In [23]:
# Checking results
df_customers.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   No of dependants Marital Status  Income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

## Quality and Consistency Checks

In [24]:
# Checking the datatypes for all columns
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   user_id           206209 non-null  int64 
 1   First Name        194950 non-null  object
 2   Last Name         206209 non-null  object
 3   Gender            206209 non-null  object
 4   State             206209 non-null  object
 5   Age               206209 non-null  int64 
 6   Date Joined       206209 non-null  object
 7   No of dependants  206209 non-null  int64 
 8   Marital Status    206209 non-null  object
 9   Income            206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [25]:
df_ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 23 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int64  
 1   user_id                int64  
 2   order_number           int64  
 3   order_day_of_week      int64  
 4   order_hour_of_day      int64  
 5   days_since_last_order  float64
 6   product_id             int64  
 7   add_to_cart_order      int64  
 8   reordered              int64  
 9   product_name           object 
 10  aisle_id               int64  
 11  department_id          int64  
 12  prices                 float64
 13  price_range_loc        object 
 14  busiest_day_of_week    object 
 15  busiest_days           object 
 16  busiest_period_of_day  object 
 17  max_order              int64  
 18  loyalty_flag           object 
 19  avg_price              float64
 20  spending_flag          object 
 21  median_days            float64
 22  order_frequency_

In [26]:
# Checking for missing/duplicate values
df_customers.isnull().sum()

user_id                 0
First Name          11259
Last Name               0
Gender                  0
State                   0
Age                     0
Date Joined             0
No of dependants        0
Marital Status          0
Income                  0
dtype: int64

In [27]:
# There seems to be some missing rows in the first name column in the df_customers dataframe
# To confirm
df_customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

This has confirmed that there are 11,259 missing values in the first name column. We do not need to make changes to this column or replace the missing values as the column is likely irrelevant for this analysis.

In [56]:
# For duplicate values
df_dups = df_customers[df_customers.duplicated()]

In [57]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, Date Joined, No of dependants, Marital Status, Income]
Index: []

There are no duplicate values

In [36]:
df_ords_prods.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_last_order    2076096
product_id                     0
add_to_cart_order              0
reordered                      0
product_name                   0
aisle_id                       0
department_id                  0
prices                         0
price_range_loc                0
busiest_day_of_week            0
busiest_days                   0
busiest_period_of_day          0
max_order                      0
loyalty_flag                   0
avg_price                      0
spending_flag                  0
median_days                    5
order_frequency_flag           0
dtype: int64

In [37]:
# Replacing NaN values in the days_since_last_order column with 0
df_ords_prods['days_since_last_order'].value_counts(dropna = False)

7.0     3476198
30.0    3067374
6.0     2517498
5.0     2124434
4.0     2078648
NaN     2076096
8.0     1931945
3.0     1876188
2.0     1463724
9.0     1217841
14.0    1029656
10.0     967187
1.0      940474
13.0     851114
11.0     820761
12.0     774164
15.0     671458
16.0     461335
0.0      448329
21.0     444142
17.0     382114
20.0     379732
18.0     350587
19.0     334882
22.0     308849
28.0     253151
23.0     225834
27.0     204061
24.0     193421
25.0     180417
26.0     177775
29.0     175470
Name: days_since_last_order, dtype: int64

In [38]:
df_ords_prods['days_since_last_order'].fillna(0, inplace = True)

In [39]:
# To confirm
df_ords_prods['days_since_last_order'].value_counts(dropna = False)

7.0     3476198
30.0    3067374
0.0     2524425
6.0     2517498
5.0     2124434
4.0     2078648
8.0     1931945
3.0     1876188
2.0     1463724
9.0     1217841
14.0    1029656
10.0     967187
1.0      940474
13.0     851114
11.0     820761
12.0     774164
15.0     671458
16.0     461335
21.0     444142
17.0     382114
20.0     379732
18.0     350587
19.0     334882
22.0     308849
28.0     253151
23.0     225834
27.0     204061
24.0     193421
25.0     180417
26.0     177775
29.0     175470
Name: days_since_last_order, dtype: int64

All the rows in the days_since_last_order have now been filled.
There are still missing values under median days.

In [47]:
# Replacing the missing values under median days with the mean value
df_ords_prods['median_days'].fillna(1.0, inplace = True)

There are now no missing values in the ords_prods dataset

## ## Merging df_customers with df_ords_prods

In [50]:
# checking the columns in both dataframes to discover which columns are similar
df_ords_prods.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...    price_range_loc  busiest_day_of_week  \
0         Soda  ...  Mid-range product       Regularly busy   
1         Soda  ...  Mid-range product       Regularly busy   
2         Soda  ...  Mid-range product       Regularly busy   
3         Soda  ...  Mid-range product           Least busy   
4         Soda  ...  Mid-range product           Least busy   

         busiest_days busiest_period_of_day max_order  loyalty_flag avg_price  \
0  Regularly busy day        Average orders        10  New customer  6.367797   
1     Least busy days        Average orders        10  New customer  6.367797   
2     Least busy days           Most orders        10  New customer  6.367797   
3     Least busy days        Average orders        10  New customer  6.367797   
4     Least busy days           Most orders        10  New customer  6.367797   

   spending_flag median_days   order_frequency_flag  
0    Low spender        20.5  Non-frequent customer  
1    Low spender        20.5  Non-frequent customer  
2    Low spender        20.5  Non-frequent customer  
3    Low spender        20.5  Non-frequent customer  
4    Low spender        20.5  Non-frequent customer  

[5 rows x 23 columns]

In [59]:
df_customers.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   No of dependants Marital Status  Income  
0                 3        married  165665  
1                 0         single   59285  
2                 2        married   99568  
3                 0         single   42049  
4                 1        married   40374

In [60]:
# Merging both dataframes
df_merged = df_ords_prods.merge(df_customers, on = 'user_id')

In [61]:
# Checking dimensions
df_merged.shape

(32404859, 32)

In [63]:
df_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    0.0         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...   order_frequency_flag  First Name  Last Name  Gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     State Age Date Joined  No of dependants Marital Status  Income  
0  Alabama  31   2/17/2019                 3        married   40423  
1  Alabama  31   2/17/2019                 3        married   40423  
2  Alabama  31   2/17/2019                 3        married   40423  
3  Alabama  31   2/17/2019                 3        married   40423  
4  Alabama  31   2/17/2019                 3        married   40423  

[5 rows x 32 columns]

Dataframes have been merged successfully

## Exporting as a pickle file

In [64]:
# Exporting merged dataframe to pickle file
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))